lib/segment/src/index/field_index/full_text_index/posting_list.rs

```diff
@@ -37,23 +40,323 @@ impl PostingList {
+            decompressed_chunk_start_index: 0,
+            #[cfg(test)]
+            last_checked: None,
+        }
+    }
+
+    // Check if the next value is in the compressed posting list.
+    // This function reuses the decompressed chunk to avoid unnecessary decompression.
+    // It is useful when the visitor is used to check the values in the increasing order.
+    pub fn contains_next(&mut self, val: &PointOffsetType) -> bool {
+        #[cfg(test)]
+        {
+            // check if the checked values are in the increasing order
+            if let Some(last_checked) = self.last_checked {
+                assert!(*val > last_checked);
+            }
+            self.last_checked = Some(*val);
+        }
+
+        if !self.postings.is_in_postings_range(*val) {
+            return false;
+        }
+
+        // check if current decompressed chunks range contains the value
+        if self.decompressed_chunk_idx.is_some() {
+            // check if value is in decompressed chunk range
+            // check for max value in the chunk only because we already checked for min value while decompression
+            let last_decompressed = &self.decompressed_chunk[BitPackerImpl::BLOCK_LEN - 1];
+            match val.cmp(last_decompressed) {
+                std::cmp::Ordering::Less => {
+                    // value is less than the last decompressed value
+                    return self.find_in_decompressed_chunk(val);
+                }
+                std::cmp::Ordering::Equal => {
+                    // value is equal to the last decompressed value
+                    return true;
+                }
+                std::cmp::Ordering::Greater => {}
+            }
+        }
+
+        // decompressed chunk is not in the range, so we need to decompress another chunk
+        // first, check if there is a chunk that contains the value
+        let chunk_index = match self.postings.find_chunk(val, self.decompressed_chunk_idx) {
+            Some(idx) => idx,
+            None => {
+                // value is in the noncompressed postings range
+                self.decompressed_chunk_idx = None;
+                return self.postings.reminder_postings.binary_search(val).is_ok();
+            }
+        };
+        // if the value is the initial value of the chunk, we don't need to decompress the chunk
+        if self.postings.chunks[chunk_index].initial == *val {
+            return true;
+        }
+
+        // second, decompress the chunk and check if the value is in the decompressed chunk
+        self.postings
+            .decompress_chunk(&self.bitpacker, chunk_index, &mut self.decompressed_chunk);
+        self.decompressed_chunk_idx = Some(chunk_index);
+        self.decompressed_chunk_start_index = 0;
+
+        // check if the value is in the decompressed chunk
+        self.find_in_decompressed_chunk(val)
+    }
+
+    fn find_in_decompressed_chunk(&mut self, val: &PointOffsetType) -> bool {
```

Comment:
It's not a method of postings list. It's a method for help structure for intersections. It's mutable because we do binary search inside not from 0. We do it from index, which we found from previous `contains_next` check

Reply:
But I agree that it's incorrect naming. I have to mark in function name that I change internal state of help stucture